### PR TITLE
fix: add debug info to thickness calculators

### LINKED ISSUE #146: [Feature Request] - add debug information to the thickness calculators
### 💻

- [X] Would you like to work on this feature?

### Feature Description

- [ ] - add lines object to the thickness calculator class (lines that are projected)
- [ ] - store dips as well
- [ ] - keep as a geopandas
- [ ] - object to be names as "lines"

### Current Situation

there's no debug information for thickness calculators

### Version

m2l  - 3.1.12 
lpf - 0.1.13

### Proposed Solution

na

### Additional Context

_No response_

### Affected Areas

- [ ] input data
- [ ] project creation
- [ ] samplers
- [ ] sorters
- [ ] stratigraphic column
- [ ] data types
- [X] Other

### LINKED ISSUE #147: [Feature Request] - thickness calculators option - max length projected lines
### 💻

- [x] Would you like to work on this feature?

### Feature Description

- Add user defined option to thickness calculators for max length of the projected lines from structures
- option allows user to define this globally (for all units) or per unit

### Current Situation

NA

### Version

_No response_

### Proposed Solution

na

### Additional Context

_No response_

### Affected Areas

- [ ] input data
- [ ] project creation
- [ ] samplers
- [ ] sorters
- [ ] stratigraphic column
- [ ] data types
- [X] Other

### PR DESCRIPTION
**Pull request by AngRodrigues**
Fixes #146 
Fixes #147. 

- Added object `lines` to each calculator -> gpd df containing linestrings + dip value(s) for each line. 
- Added warning if more than 75% of the units don't have a calculated thickness
- Added option for max_line_length

If simple review, we could add this to 3.2 too.

### CODE REVIEW

**Comment by AngRodrigues** [comment]
A note that I will merge this into 3.2 by COB today if no review

File: map2loop/thickness_calculator.py [hunk 1]
```diff
@@ -372,7 +393,10 @@ def compute(
                 logger.warning(
                     f"Thickness Calculator InterpolatedStructure: Cannot calculate thickness between {stratigraphic_order[i]} and {stratigraphic_order[i + 1]}\n"
                 )
-
+                
+        self.lines = geopandas.GeoDataFrame(geometry=[line[0] for line in _lines], crs = basal_contacts.crs)
```
Comment (lachlangrose) [inline map2loop/thickness_calculator.py]: Change lines to remove the list comprehension

File: map2loop/thickness_calculator.py [hunk 2]
```diff
@@ -199,6 +211,7 @@ def __init__(self):
         """
         self.thickness_calculator_label = "InterpolatedStructure"
         self.lines = []
+        self.max_line_length = None
```
Comment (lachlangrose) [inline map2loop/thickness_calculator.py]: Where is this initialised?

File: map2loop/thickness_calculator.py [hunk 3]
```diff
@@ -199,6 +211,7 @@ def __init__(self):
         """
         self.thickness_calculator_label = "InterpolatedStructure"
         self.lines = []
```
Comment (lachlangrose) [inline map2loop/thickness_calculator.py]: Add the interpolated orientations

Comment (lachlangrose) [reply map2loop/thickness_calculator.py]: Also can we keep track of the thickness estimates for the locations on the map

File: map2loop/thickness_calculator.py [hunk 4]
```diff
@@ -394,6 +418,8 @@ def __init__(self):
         self.thickness_calculator_label = "StructuralPoint"
```
Comment (lachlangrose) [inline map2loop/thickness_calculator.py:442]: Why are the attributes hard coded? line length should be set per map. Make these variables of the init function with these values as default

File: map2loop/thickness_calculator.py [hunk 5]
```diff
@@ -394,6 +418,8 @@ def __init__(self):
...
         self.line_length = 10000
         self.strike_allowance = 30
+        self.max_line_length = None
+        self.lines = []
```
Comment (lachlangrose) [inline map2loop/thickness_calculator.py]: Don't initialise lines as a list and then change it to a geopandas dataframe, make it either None or an empty list

File: map2loop/thickness_calculator.py (hunk 1 above)
Comment (AngRodrigues) [reply map2loop/thickness_calculator.py]: https://github.com/Loop3D/map2loop/pull/161/commits/63af48d7d2bbb644a0a004efc2289a07ccb90dff

Comment (AngRodrigues) [reply map2loop/thickness_calculator.py]: Had to revert this back. 
The lst comprehension was the only way I could find to build the gdf. I do not know of another way to build the gdf - please let me know if you have any suggestions about this @lachlangrose

File: map2loop/thickness_calculator.py (hunk 5 above)
Comment (AngRodrigues) [reply map2loop/thickness_calculator.py]: https://github.com/Loop3D/map2loop/pull/161/commits/6a43ee9de156823a39e5db6b097e5e8be7245904

File: map2loop/thickness_calculator.py (hunk 2 above)
Comment (AngRodrigues) [reply map2loop/thickness_calculator.py]: assuming this has to do with your comment below, so will resolve this thread for now. Please do let me know if not related. 

> Why are the attributes hard coded? line length should be set per map. Make these variables of the init function with these values as default

File: map2loop/thickness_calculator.py (hunk 4 above)
Comment (AngRodrigues) [reply map2loop/thickness_calculator.py:442]: https://github.com/Loop3D/map2loop/pull/161/commits/d47ac30847e64c69c416528e93646bbe2d95f281 & https://github.com/Loop3D/map2loop/pull/161/commits/837377cfb6d658e81a43f5b1f1294a4fb1a569ed

Max line length is now an attribute of the abstract class Default is None for all calculators. structuralPoint needs a value, which is set to 10000m as previous versions.

File: map2loop/thickness_calculator.py (hunk 3 above)
Comment (AngRodrigues) [reply map2loop/thickness_calculator.py]: Will add this as a feature request for the Interpolated structure thickness calculator.

Comment (AngRodrigues) [reply map2loop/thickness_calculator.py]: The interpolated orientations are stored as the "dip" column in the `self.lines` object.
 
Added an extra object called `self.location_tracking` that contains the points coordinates and the thickness estimates - https://github.com/Loop3D/map2loop/pull/161/commits/797053b69532cc4f26a41a296bcbdd71d82561e6

File: map2loop/thickness_calculator.py (hunk 1 above)
Comment (AngRodrigues) [reply map2loop/thickness_calculator.py]: Fixed in https://github.com/Loop3D/map2loop/pull/161/commits/bf22a328c50d4def89ad78f5e600c9b35256b1ae

**Pull request merged**